plugins/modules/digital_ocean_snapshot.py

```diff
@@ -0,0 +1,303 @@
+  community.digitalocean.digital_ocean_snapshot:
+    state: present
+    snapshot_type: volume
+    snapshot_name: mysnapshot1
+    volume_id: 9db5e329-cc68-11eb-b027-0a58ac144f91
+
+- name: Delete a Volume snapshot
+  community.digitalocean.digital_ocean_snapshot:
+    state: absent
+    snapshot_type: volume
+    snapshot_id: a902cdba-cc68-11eb-a701-0a58ac145708
+"""
+
+
+RETURN = r"""
+data:
+  description: Snapshot creation or deletion action.
+  returned: success
+  type: dict
+  sample:
+    - completed_at: '2021-06-14T12:36:00Z'
+      id: 1229119156
+      region:
+        available: true
+        features:
+          - backups
+          - ipv6
+          - metadata
+          - install_agent
+          - storage
+          - image_transfer
+        name: New York 1
+        sizes:
+          - s-1vcpu-1gb
+          - s-1vcpu-1gb-amd
+          - s-1vcpu-1gb-intel
+          - <snip>
+        slug: nyc1
+      region_slug: nyc1
+      resource_id: 250445117
+      resource_type: droplet
+      started_at: '2021-06-14T12:35:25Z'
+      status: completed
+      type: snapshot
+    - created_at: '2021-06-14T12:55:10Z'
+      id: c06d4a86-cd0f-11eb-b13c-0a58ac145472
+      min_disk_size: 1
+      name: my-snapshot-1
+      regions:
+        - nbg1
+      resource_id: f0adea7e-cd0d-11eb-b027-0a58ac144f91
+      resource_type: volume
+      size_gigabytes: 0
+      tags:
+        - tag1
+        - tag2
+"""
+
+
+import time
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.community.digitalocean.plugins.module_utils.digital_ocean import (
+    DigitalOceanHelper,
+)
+
+
+class DOSnapshot(object):
+    def __init__(self, module):
+        self.rest = DigitalOceanHelper(module)
+        self.module = module
+        self.wait = self.module.params.pop("wait", True)
+        self.wait_timeout = self.module.params.pop("wait_timeout", 120)
+        # pop the oauth token so we don't include it in the POST data
+        self.module.params.pop("oauth_token")
+        self.snapshot_type = module.params["snapshot_type"]
+        self.snapshot_name = module.params["snapshot_name"]
+        self.snapshot_tags = module.params["snapshot_tags"]
+        self.snapshot_id = module.params["snapshot_id"]
+        self.volume_id = module.params["volume_id"]
+
+    def wait_finished(self):
+        end_time = time.monotonic() + self.wait_timeout
+        while time.monotonic() < end_time:
+            response = self.rest.get("actions/{0}".format(str(self.action_id)))
+            status = response.status_code
+            if status != 200:
+                self.module.fail_json(
+                    msg="Unable to find action {0}, please file a bug".format(
+                        str(self.action_id)
+                    )
+                )
+            json = response.json
+            if json["action"]["status"] == "completed":
+                return json
+            time.sleep(min(10, end_time - time.monotonic()))
```

Comment:
It is possible for `ValueError: sleep length must be non-negative` to happen here, if by the time this line is reached `time.monotonic()` is greater than `end_time`.

Reply:
Ah yeah, the other idea is ... why are we trying to get savings here at all ... we're talking about a few seconds, heh.